### PR TITLE
Feature/document expiry for ObjC

### PR DESCRIPTION
**Pull request by jayahariv**
* Same as 1.x, when the database is open, schedule to clean the expiration docs.
* When setting the expiration, schedule to clean the expiration docs right away.
* When the database is closed or deleted, please make sure to unschedule the cleanup previously scheduled.
* Make sure to make the API thread safe.

#2227

### CODE REVIEW

File: Objective-C/CBLDatabase.mm [hunk 1]
```diff
@@ -969,4 +1015,50 @@ - (BOOL) saveResolvedDocument: (CBLDocument*)resolvedDoc
...
+# pragma mark DOCUMENT EXPIRATION
+
+
+- (void) scheduleDocumentExpiration: (NSTimeInterval)minimumDelay {
+    [self cancelPreviousPerformRequestsForPurgingExpiredDocuments];
+    
+    UInt64 nextExpiration = c4db_nextDocExpiration(_c4db);
+    if (nextExpiration > 0) {
+        NSDate* expDate = [NSDate dateWithTimeIntervalSince1970: nextExpiration];
+        NSTimeInterval delay = MAX(expDate.timeIntervalSinceNow + 1.0, minimumDelay);
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: I'm not certain if adding an extra one second is needed.

File: Objective-C/Tests/DatabaseTest.m [hunk 2]
```diff
@@ -1911,4 +1911,309 @@ - (void) testDeleteIndex {
...
+    AssertEqual(self.db.count, 1u);
+    
+    // set expiry
+    NSTimeInterval expiryTime = 3;
+    NSDate* expiryDate = [NSDate dateWithTimeIntervalSinceNow: expiryTime];
+    NSError* err;
+    Assert([self.db setDocumentExpirationWithID: docID date: expiryDate error: &err]);
+    
+    // validate result
+    Assert([expiryDate timeIntervalSinceDate:[self.db getDocumentExpirationWithID: docID]] < 1);
```
[the remark this answers is not in the document]
Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m]: No they are not equal, the date we will be supplying has precision with milliseconds. for example, 12:03:24.67(date we supply, in this case expiryDate), and the expiration time returned from the lite-core is 12:03:24.00(in this case, [self.db getDocumentExpirationWithID: docID]).

File: Objective-C/CBLDatabase.mm (hunk 1 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: I will check this.

File: Objective-C/Tests/DatabaseTest.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (pasin) [reply Objective-C/Tests/DatabaseTest.m]: I see.

File: Objective-C/CBLDatabase.mm (hunk 1 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: removed the extra buffer time, and all tests are still passing.

File: Objective-C/CBLDatabase.mm [hunk 3]
```diff
@@ -1044,7 +1044,7 @@ - (void) purgeExpiredDocuments {
         C4Error err;
         NSUInteger nPurged = c4db_purgeExpiredDocs(_c4db, &err);
         CBLLog(Database, @"Purged %lu expired documents", (unsigned long)nPurged);
-        [self scheduleDocumentExpiration: 1.0];
+        [self scheduleDocumentExpiration: 0.0];
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: Please add one second back to this for now as we want to have some interval between checking the expired doc to purge. We can discuss an appropriate interval on Friday.

File: Objective-C/CBLDatabase.h [hunk 4]
```diff
@@ -394,6 +394,29 @@ typedef NS_ENUM(uint32_t, CBLConcurrencyControl) {
...
+ Sets an expiration date on a document.
+ After this time the document will be purged from the database.
+ 
+ @param documentID The ID of the document to set the expiration date for
+ @param date The expiration date. Set nil date will reset the document expiration.
+ @param error error On return, the error if any.
+ @return True on success, false on failure.
+ */
+- (BOOL) setDocumentExpirationWithID: (NSString*)documentID
+                                date: (nullable NSDate*)date
```
Comment (pasin) [inline Objective-C/CBLDatabase.h]: Can you change date to expiration?

**Pull request merged**